File: examples/python/flask/web.py
import uuid
from datetime import datetime
from typing import Optional

from fief_client import Fief, FiefUserInfo
from fief_client.integrations.flask import (
    FiefAuth,
    FiefAuthForbidden,
    FiefAuthUnauthorized,
    get_cookie,
)
from flask import Flask, g, redirect, request, session, url_for

SESSION_COOKIE_NAME = "user_session"
SECRET_KEY = "SECRET"  # (1)!


fief = Fief(  # (2)!
    "https://example.fief.dev",
    "YOUR_CLIENT_ID",
    "YOUR_CLIENT_SECRET",
)


def get_userinfo_cache(id: uuid.UUID) -> Optional[FiefUserInfo]:  # (3)!
    return session.get(f"userinfo-{str(id)}")


def set_userinfo_cache(id: uuid.UUID, userinfo: FiefUserInfo) -> None:  # (4)!
    session[f"userinfo-{str(id)}"] = userinfo


auth = FiefAuth(
    fief,
    get_cookie(SESSION_COOKIE_NAME),  # (5)
    get_userinfo_cache=get_userinfo_cache,  # (6)!
    set_userinfo_cache=set_userinfo_cache,
)
app = Flask(__name__)
app.secret_key = SECRET_KEY


@app.errorhandler(FiefAuthUnauthorized)  # (7)!
def fief_unauthorized_error(e):
    redirect_uri = url_for("auth_callback", _external=True)  # (8)!
    auth_url = fief.auth_url(redirect_uri, scope=["openid"])  # (9)!
    return redirect(auth_url)  # (10)!


@app.errorhandler(FiefAuthForbidden)
def fief_forbidden_error(e):
    return "", 403


@app.get("/auth-callback")  # (11)!
def auth_callback():
    redirect_uri = url_for("auth_callback", _external=True)
    code = request.args["code"]
    tokens, userinfo = fief.auth_callback(code, redirect_uri)  # (12)!

    response = redirect(url_for("protected"))  # (13)!
    response.set_cookie(  # (14)!
        SESSION_COOKIE_NAME,
        tokens["access_token"],
        expires=int(datetime.now().timestamp() + tokens["expires_in"]),
        httponly=True,  # (15)!
        secure=False,  # ❌ Set this to `True` in production (16)!
    )

    set_userinfo_cache(uuid.UUID(userinfo["sub"]), userinfo)  # (17)!

    return response


@app.get("/protected")
@auth.current_user()  # (18)!
def protected():
    user = g.user  # (19)!
    return f"<h1>You are authenticated. Your user email is {user['email']}</h1>"

File: examples/javascript/express/web.js
const fief = require('@fief/fief');
const fiefExpress = require('@fief/fief/express');
const express = require('express');

const SESSION_COOKIE_NAME = "user_session"  // (1)!
const PORT = 3000;
const REDIRECT_URI = `http://localhost:${PORT}/auth-callback`;  // (2)!

class MemoryUserInfoCache {  // (3)!
  constructor() {
    this.storage = {};
  }

  async get(id) {
    const userinfo = this.storage[id];
    if (userinfo) {
      return userinfo;
    }
    return null;
  }

  async set(id, userinfo) {
    this.storage[id] = userinfo;
  }

  async remove() {
    this.storage[id] = undefined;
  }

  async clear() {
    this.storage = {};
  }
}

const userInfoCache = new MemoryUserInfoCache();

const fiefClient = new fief.Fief({  // (4)!
  baseURL: 'https://example.fief.dev',
  clientId: 'YOUR_CLIENT_ID',
  clientSecret: 'YOUR_CLIENT_SECRET',
});

const unauthorizedResponse = async (req, res) => {  // (5)!
  const authURL = await fiefClient.getAuthURL({ redirectURI: REDIRECT_URI, scope: ['openid'] });  // (6)!
  res.redirect(307, authURL);  // (7)!
};

const fiefAuthMiddleware = fiefExpress.createMiddleware({
  client: fiefClient,
  tokenGetter: fiefExpress.cookieGetter(SESSION_COOKIE_NAME),  // (8)!
  unauthorizedResponse,  // (9)!
  userInfoCache,  // (10)!
});

const app = express();

app.get('/auth-callback', async (req, res) => {  // (11)!
  const code = req.query['code'];
  const [tokens, userinfo] = await fiefClient.authCallback(code, REDIRECT_URI);  // (12)!

  userInfoCache.set(userinfo.sub, userinfo);  // (13)!

  res.cookie(  // (14)!
    SESSION_COOKIE_NAME,
    tokens.access_token,
    {
      maxAge: tokens.expires_in * 1000,
      httpOnly: true,  // (15)!
      secure: false,  // ❌ Set this to `true` in production (16)!
    },
  );
  res.redirect('/protected');
});

app.get('/protected', fiefAuthMiddleware(), (req, res) => {  // (17)!
  res.send(`<h1>You are authenticated. Your user email is ${req.user.email}</h1>`)  // (18)!
});

app.listen(PORT, () => {
  console.log(`Example app listening on port ${PORT}`)
});
